packages/teuchos/core/src/Teuchos_RCPNode.hpp

```diff
@@ -249,8 +299,17 @@ class TEUCHOSCORE_LIB_DLL_EXPORT RCPNode {
     EPrePostDestruction destroy_when;
   };
   typedef Teuchos::map<std::string,extra_data_entry_t> extra_data_map_t;
-  int count_[2];
+
+#ifdef USING_ATOMICS
```

Comment:
To reduce duplication, you want to replace this with:

```
#ifdef USE_ATOMICS
#  define TEUCHOS_RCP_DECL_ATOMIC(VAR, T) std::atomic<T> VAR
#else
#  define TEUCHOS_RCP_DECL_ATOMIC(VAR, T) T VAR
#endif

  TEUCHOS_RCP_DECL_ATOMIC(strong_count_, int);
  TEUCHOS_RCP_DECL_ATOMIC(weak_count_plus_, int);
  TEUCHOS_RCP_DECL_ATOMIC(has_ownership_, int);
```

You should move the the define for TEUCHOS_RCP_DECL_ATOMIC() out of the class del to avoid clutter.  You might also want to reuse this in other places perhaps.